Add methods for read/write csv to enforce standards

anthonysena opened the pull request:
Per #26, this set of changes aims to add exported functions for reading/writing CSV files so that other HADES packages have a consistent mechanism for these operations. Additionally, these functions aim to enforce the rules described in #26. 

At the moment, the `writeCsv` function provides warnings as opposed to errors. My thinking is that a warning would alert the user to the potential problem if the file is intended for upload but it should not stop you from actually writing the file. Conversely, if you are using `writeCsv` with the intent of storing the results to upload to a database, shouldn't we stop the file writing all together?

I'm unsure how to best handle this so tagging @schuemie for any feedback here.

anthonysena commented on R/CsvHelper.R:
```diff
@@ -0,0 +1,126 @@
+    # Also perform a check to see if the fileName end in "s" which might indicate
+    # that the resulting file name is plural
+    if (endsWith(fileName, "s")) {
+      message(paste("The filename:", basename(file), "may be plural since it ends in 's'. Please ensure you are using singular nouns for your file names."))
```
I've made this a `message` as opposed to a warning to alert the user.

anthonysena commented on R/CsvHelper.R:
```diff
@@ -0,0 +1,126 @@
+#'  Returns the input x invisibly.
+#'
+#' @export
+writeCsv <- function(x, file, warnOnUploadRuleViolations = TRUE) {
```
I'm wondering if I should change the default of `warnOnUploadRuleViolations` to FALSE so that we only raise warnings when the user wants to know if they may be violating rules?

schuemie commented on R/CsvHelper.R:
```diff
@@ -0,0 +1,125 @@
+#'
+#' @export
+isSnakeCase <- function(x) {
+  y <- snakecase::to_snake_case(x, numerals = "asis")
```
I would add an entire package dependency just to check if a string is snake case. A simple regex could do that, right?

anthonysena replied:
Yes indeed. I'm also using the `snakecase` package to suggest how the column names could be adjusted to use snakecase. That package offered some additional benefits in terms of automatically preserving names that were already in snakecase, etc. I'm not married to using it but thought it was potentially useful.

anthonysena replied:
I've removed the dependency on snakecase and opted for the regex approach.

schuemie commented:
Looks good, although I would probably assume that the variable names in the data frames are camel case as defined in the [HADES style guide](https://ohdsi.github.io/Hades/codeStyle.html#Case), and convert to snake case on the fly. 

In general, I would not add an entire dependency just for the snake case. There are conversion functions in SqlRender we could use, and we could easily add functions for detecting snake case (or camel case) there.

anthonysena commented:
> Looks good, although I would probably assume that the variable names in the data frames are camel case as defined in the [HADES style guide](https://ohdsi.github.io/Hades/codeStyle.html#Case), and convert to snake case on the fly.

I had originally thought to automatically switch from camelCase to snakeCase inside of the function but I was unsure if I wanted to impose that on the caller? We could also consider having another set of functions for reading/writing results `readResultsCsv` and `writeResultsCsv` which impose these rules?

anthonysena commented:
@schuemie based on your feedback I've revised this PR with these changes: https://github.com/OHDSI/CohortGenerator/pull/27/commits/76ac1a2855cc870ca6186f0ddc7d66bac2f1e9ad. Specifically:

- I've removed the `snakecase` package dependency and opted for using regex for checking the case formatting. We could consider moving these to SqlRender later if you agree.
- `readCsv` and `writeCsv` now have a parameter `warnOnCaseMismatch` that will provide a warning if your input is not formatted properly. When `warnOnCaseMismatch == TRUE`, readCsv will check to see if the columns in the .csv file are in snake_case format. When `warnOnCaseMismatch == TRUE`, writeCsv will check to see if the data.frame columns are in camelCase format. 
- I've updated and expanded the test cases related to these changes.

From my perspective, these changes are now ready to squash and merge into develop. Let me know if you agree and I can close this out. Thanks!

anthonysena commented on R/Export.R:
```diff
@@ -59,17 +59,17 @@ exportCohortStatsTables <- function(connectionDetails,
     data <- getStatsTable(connection = connection,
                           table = table,
-                          snakeCaseToCamelCase = FALSE,
+                          snakeCaseToCamelCase = TRUE,
```
Need to revise this to allow for toggling of snake case per #23
